### PR TITLE
Implements an `OLMo-core` compatible data loader to load HF datasets from `get_cached_dataset_tulu`. 

### PR DESCRIPTION
**Pull request by finbarrtimbers**
There's not a ton of point to this, right now. But by making this change, we can subsequently: 

1) move `dpo_tune_cache.py` over to use this, which will (eventually) let that use the Olmo-core trainer
2) Give each LLMRayActor their own data_loader, so they can pull independently and not have to coordinate with `main`. This will let us remove `add_prompt_to_generator` and have them operate fully asynchronously. This will make switching the entire `grpo_fast.py` script over to the data_loader/trainer pattern easier.

Runs: 

- Single GPU GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KAC7FADGWJFMVMF7Z6A7MHWK?taskId=01KAC7FADMXYSEG1C5S26VMPDE&jobId=01KAC7FANAP7ECR5VH70VN7T0F)
- Multi-node GRPO script: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KAVP00AZWTA505ZGRND8QSKV?taskId=01KAVP00B2RDE2ACT982ZRD0H6&jobId=01KAVP00EHXXCT51MAWN6PSX2R)

---

> [!NOTE]
> Introduces an OLMo-core–compatible HF dataset loader and refactors GRPO to use it (replacing custom iterators/maps) with prompt_id-based request tracking; updates tests and dependencies.
> 
> - **Data Loading**:
>   - Add `HFDataLoader` (`open_instruct/data_loader.py`) implementing `DataLoaderBase` with HF `Dataset` sharding, shuffling, exclusion, checkpointing (`state_dict`/`load_state_dict`), optional auto-reshuffle, and per-example `prompt_id`.
> - **GRPO Pipeline Refactor** (`open_instruct/grpo_fast.py`):
>   - Remove `ShufflingIterator` and `PendingQueriesMap`; use `HFDataLoader` for prompt sourcing, replenishment, and no-resampling via `exclude_index`.
>   - Update `accumulate_inference_batches`, `data_preparation_thread`, `run_training`, and eval flow to consume from `HFDataLoader` and lookup examples by `dataset_index`.
>   - Simplify `add_prompt_to_generator(example, param_prompt_Q, generation_config, is_eval)`; batch/eval queuing now uses `prompt_id`.
>   - Checkpoint now saves/loads `dataloader_state` instead of iterator state.
> - **Queue/ID Schema** (`open_instruct/queue_types.py`, `open_instruct/vllm_utils.py`):
>   - `PromptRequest`/`GenerationResult` now carry `dataset_index` and `prompt_id` (drop `epoch_number`/`training_step`).
>   - Request ID format changed to `{train|eval}_<prompt_id>`; metadata/plumbing updated accordingly.
> - **Tests**:
>   - Add `test_data_loader.py` covering sharding, reshuffle, checkpointing, exclusion, and `prompt_id` uniqueness.
>   - Update/refactor `test_grpo_fast.py` and `test_vllm_utils.py` for new dataloader and `prompt_id` APIs.
> - **Dependencies**:
>   - Add `ai2-olmo-core==2.3.0` and related deps; lockfile updates.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b523afa9a6a557ddfd11629eb5622e888d3f8af6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/data_loader.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    def _iter_batches(self) -> Iterable[dict[str, Any]]:
+        """Return an iterable over all batches in the epoch."""
+        epoch = self._epoch or 0
+        for i in range(self.batches_processed, self.effective_size):
+            example = self.dataset[i]
+            yield example | {"prompt_id": f"{epoch}_{example['dataset_index']}"}
+
+    @property
+    def total_batches(self) -> int:
+        """Return the total number of batches in an epoch."""
```
Comment (hamishivi) [inline open_instruct/data_loader.py:77]: Is this doc string right? Isn't effective size number of samples in the dataset, from line 38?

File: open_instruct/grpo_fast.py [hunk 2]
```diff
@@ -1817,10 +1678,9 @@ def accumulate_inference_batches(
         percent_solved = np.mean(scores).item() / args.max_possible_score
         # Don't resample prompt that was solved at more than no_resample_positive_rate
         if no_resampling_pass_rate is not None and percent_solved >= no_resampling_pass_rate:
-            iter_dataloader.exclude_index(result.dataset_index)
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py:1823]: was this change intentional?

File: open_instruct/grpo_fast.py [hunk 3]
```diff
@@ -1934,8 +1794,8 @@ def accumulate_inference_batches(
         finish_reasons=combined_finish_reasons,
         masks=combined_masks,
         request_info=combined_request_info,
-        dataset_index=None,
-        epoch_number=results[0].epoch_number,
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py:1941]: intentional change...?

File: open_instruct/grpo_fast.py [hunk 4]
```diff
@@ -3025,6 +2875,14 @@ def run_training(
     if resume_training_step > 1:
         logger.info(f"[Main Thread] Resuming training from step {resume_training_step}")
 
+    data_loader_iter = iter(data_loader)
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py]: Do we need to call reshuffle at some point for multi epoch training?

File: open_instruct/grpo_fast.py [hunk 5]
```diff
@@ -3072,17 +2929,8 @@ def health_check_fn():
...
-            train_dataset[dataset_index],
-            dataset_index,
-            iter_dataloader.epoch_number,
-            resume_training_step,
-            pending_queries_map,
-            param_prompt_Q,
-            generation_configs["train"],
-            is_eval=False,
-        )
+        example = next(data_loader_iter)
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py]: usual question here: how do we want/did we decide to handle the case where the async steps go over an epoch boundary?

File: open_instruct/grpo_fast.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py:1823]: Mistake! Fixed.

File: open_instruct/data_loader.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/data_loader.py:77]: yes, you're right. Fixed!

File: open_instruct/grpo_fast.py (hunk 3 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py:1941]: Yes, this is. It's tough to get dataset_index + epoch_number from the current data_loader API, so instead, I'm having it return a `request_id` parameter in the dataset, which is unique. It's `"{epoch_number}_{dataset_index}"`.

File: open_instruct/grpo_fast.py (hunk 5 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py]: Ah, good point. I'll copy the same behaviour as ShufflingIterator.

File: open_instruct/grpo_fast.py (hunk 4 above)
Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py]: I've fixed this to have a flag, `automatic_reshuffle`, which, when True, matches the current behaviour of `ShufflingIterator`.

File: open_instruct/grpo_fast.py [hunk 6]
```diff
@@ -275,6 +276,8 @@ class Args:
     """Whether to filter out prompts with zero reward std (all samples have the same score)."""
     no_resampling_pass_rate: float | None = None
     """If the response to a prompt is solved at a rate higher than this, do not resample this prompt again"""
+    automatic_reshuffle: bool = True
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py]: If this flag is false, then training just errors right? is there a reason to have it configurable?

Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py]: Great point! Removed.

Comment (hamishivi) [reply open_instruct/grpo_fast.py]: Not sure that this got removed! - we dont need this flag on the user end, I think.

**Comment by finbarrtimbers** [comment]
Finally removed the flag! Whoops. I now set `automatic_reshuffle=True` for the train data_loader, and False for eval.

**Comment by finbarrtimbers** [comment]
Re-ran experiments: 

1. Single GPU GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBEMP16B9VNRRR27FGSADXX7?)
2. Multi-node GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBEMQ024XVZATJ1Y63ACW96G?)
